### PR TITLE
WIP: Dummy gc

### PR DESCRIPTION
**Pull request by Ngoguey42**
The expected steps for this:
image

### CODE REVIEW

**Comment by icristescu** [comment]
I added here https://github.com/mirage/irmin/pull/1905/commits a few commits that add the magic 'Z' in pack values and fix the reload.

File: src/irmin-pack/inode_intf.ml [hunk 1]
```diff
@@ -85,6 +85,8 @@ module type S = sig
   include Irmin.Indexable.S
   module Hash : Irmin.Hash.S with type t = hash
 
+  val unsafe_find : check_integrity:bool -> [< read ] t -> key -> value option
```
Comment (Ngoguey42) [inline src/irmin-pack/inode_intf.ml:88]: May be merged separately

File: src/irmin-pack/unix/append_only_file_intf.ml [hunk 2]
```diff
@@ -151,6 +151,9 @@ module type S = sig
       Always returns [Error `Rw_not_allowed]. *)
 
   val readonly : t -> bool
+  val auto_flush_threshold : t -> int option
+  val pending_flush : t -> bool
+  val path : t -> string
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/append_only_file_intf.ml:156]: May be merged separately

File: src/irmin-pack/unix/append_only_file.ml [hunk 3]
```diff
@@ -113,13 +113,21 @@ module Make (Io : Io.S) = struct
...
   let read_to_string t ~off ~len =
     let ( + ) = Int63.add in
-    let off = off + t.dead_header_size in
-    Io.read_to_string t.io ~off ~len
+    let ( > ) a b = Int63.compare a b > 0 in
+    let off' = off + Int63.of_int len in
+    if off' > t.persisted_end_offset then Error `Read_out_of_bounds
+    else
+      let off = off + t.dead_header_size in
+      Io.read_to_string t.io ~off ~len
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/append_only_file.ml:130]: Should be ported to #1865

File: src/irmin-pack/unix/errors.ml [hunk 4]
```diff
@@ -17,28 +17,39 @@
 include Errors_base
 
 module type S = sig
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/errors.ml:19]: Diffs in `errors` and `errors_base` may be merged separately

File: src/irmin-pack/unix/ext.ml [hunk 5]
```diff
@@ -134,25 +148,30 @@ module Maker (Config : Conf.S) = struct
           branch : Branch.t;
           fm : File_manager.t;
           dict : Dict.t;
+          mutable during_batch : bool;
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/ext.ml:151]: May be merged separately

File: src/irmin-pack/unix/file_manager.ml [hunk 6]
```diff
@@ -106,53 +108,66 @@ struct
...
-    let* () = Dict.flush t.dict in
-    let* () = if t.use_fsync then Dict.fsync t.dict else Ok () in
-    let* () =
-      let pl : Payload.t = Control.payload t.control in
-      let pl = { pl with dict_offset_end = Dict.end_offset t.dict } in
-      Control.set_payload t.control pl
-    in
-    let+ () = if t.use_fsync then Control.fsync t.control else Ok () in
-    ()
+    if not @@ Dict.pending_flush t.dict then Ok ()
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/file_manager.ml:118]: Should be ported to #1865

File: src/irmin-pack/unix/file_manager.ml [hunk 7]
```diff
@@ -106,53 +108,66 @@ struct
...
-                 possible to GC this irmin-pack store anymore."];
-              Payload.From_v3_used_non_minimal_indexing_strategy)
-        | From_v3_used_non_minimal_indexing_strategy -> pl.status
-        | T0 | T1 | T2 | T3 | T4 | T5 | T6 | T7 | T8 | T9 | T10 | T11 | T12
-        | T13 | T14 | T15 ->
-            assert false
-      in
-      let pl =
-        { pl with entry_offset_suffix_end = Suffix.end_offset t.suffix; status }
+    if not @@ Suffix.pending_flush t.suffix then Ok ()
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/file_manager.ml:134]: Should be ported to #1865

File: src/irmin-pack/unix/io.ml [hunk 8]
```diff
@@ -124,7 +127,9 @@ module Unix = struct
               in
               Ok { fd; closed = false; readonly = false; path }
           | false -> Error (`File_exists path))
-    with Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+    with
+    | Unix.Unix_error (e, s1, s2) -> Error (`Io_misc (e, s1, s2))
+    | Sys_error _ -> assert false
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/io.ml:132]: Diffs in `io` and `io_intf` must be ported to #1865

File: src/irmin-pack/unix/pack_store_intf.ml [hunk 9]
```diff
@@ -11,9 +11,7 @@ module type S = sig
   type file_manager
   type dict
 
-  val v :
-    config:Irmin.Backend.Conf.t -> fm:file_manager -> dict:dict -> read t Lwt.t
-
+  val v : config:Irmin.Backend.Conf.t -> fm:file_manager -> dict:dict -> read t
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/pack_store_intf.ml:14]: May be merged separately

File: src/irmin-pack/unix/pack_store_intf.ml [hunk 10]
```diff
@@ -40,6 +38,8 @@ module type S = sig
 
   val index_direct_with_kind : 'a t -> hash -> (key * Pack_value.Kind.t) option
   (** Returns the key and the kind of an object indexed by hash. *)
+
+  val purge_lru : 'a t -> unit
```
Comment (Ngoguey42) [inline src/irmin-pack/unix/pack_store_intf.ml:42]: May be merged separately

**Pull request closed without merging**